core/src/main/java/org/elasticsearch/node/Node.java

```diff
@@ -282,28 +289,73 @@ public Node start() {
         injector.getInstance(MonitorService.class).start();
         injector.getInstance(RestController.class).start();
 
+        assert injector.getInstance(ClusterService.class) instanceof InternalClusterService :
+                "node cluster service implementation must inherit from InternalClusterService";
+        final InternalClusterService clusterService = (InternalClusterService) injector.getInstance(ClusterService.class);
+
         // TODO hack around circular dependencies problems
-        injector.getInstance(GatewayAllocator.class).setReallocation(injector.getInstance(ClusterService.class), injector.getInstance(RoutingService.class));
+        injector.getInstance(GatewayAllocator.class).setReallocation(clusterService, injector.getInstance(RoutingService.class));
 
         injector.getInstance(ResourceWatcherService.class).start();
         injector.getInstance(GatewayService.class).start();
+        Discovery discovery = injector.getInstance(Discovery.class);
+        clusterService.addInitialStateBlock(discovery.getDiscoverySettings().getNoMasterBlock());
+        clusterService.setClusterStatePublisher(discovery::publish);
+
+        // start before the cluster service since it adds/removes initial Cluster state blocks
+        final TribeService tribeService = injector.getInstance(TribeService.class);
+        tribeService.start();
```

Comment:
Can we explicitly set the blocks here? Advantage is that
- no need for TribeService to depend on DiscoveryService
- no need for newly introduced method `removeInitialStateBlock(int id)` as we know exactly which blocks we previously applied.

Even better would be to also set the STATE_NOT_RECOVERED_BLOCK block for GatewayService here.

We could then not set these blocks in the first place if `tribeService.getNodes().isEmpty() == false`.